libpromises/evalfunction.c

```diff
@@ -6180,106 +6180,119 @@ static FnCallResult FnCallReadRealList(EvalContext *ctx, ARG_UNUSED const Policy
-        }
-    }
-    else
+    if (strcmp("CSV", requested_mode) == 0 ||
+        strcmp("ENV", requested_mode) == 0)
     {
-        size_max = args->next ? IntFromString(RlistScalarValue(args->next)) : IntFromString("inf");
-
-        if (strcmp(fp->name, "readyaml") == 0)
-        {
-            requested_mode = "YAML";
-        }
-        else if (strcmp(fp->name, "readcsv") == 0)
-        {
-            requested_mode = "CSV";
-        }
-        else
-        {
-            requested_mode = "JSON";
-        }
-    }
+        bool env_mode = (strcmp("ENV", requested_mode) == 0);
 
-    if (strcmp("CSV", requested_mode) == 0)
-    {
         size_t size_max = 50 * (1024 * 1024);
 
         size_t byte_count = 0;
 
         FILE *fin = safe_fopen(input_path, "r");
         if (fin == NULL)
         {
-            Log(LOG_LEVEL_VERBOSE, "%s cannot open the CSV file '%s' (fopen: %s)",
-                fp->name, input_path, GetErrorStr());
-            return FnFailure();
+            Log(LOG_LEVEL_VERBOSE, "%s cannot open the %s file '%s' (fopen: %s)",
+                myname, requested_mode, input_path, GetErrorStr());
+            return NULL;
         }
 
-        JsonElement *json = JsonArrayCreate(50);
+        JsonElement *json = NULL;
         int linenumber = 0;
-        char *line;
-        while ((line = GetCsvLineNext(fin)) != NULL)
+        if (env_mode)
         {
-            ++linenumber;
+            json = JsonObjectCreate(10);
+            size_t line_size = CF_BUFSIZE;
+            char *line = xmalloc(CF_BUFSIZE);
+
 
-            byte_count += strlen(line);
-            if (byte_count > size_max)
+            while (CfReadLine(&line, &line_size, fin) != -1)
             {
-                Log(LOG_LEVEL_VERBOSE, "%s: line %d from CSV file '%s' exceeded byte limit %lu, done with file",
-                    fp->name, linenumber, input_path, (long unsigned int)size_max);
-                free(line);
-                break;
+                ++linenumber;
+
+                byte_count += strlen(line);
+                if (byte_count > size_max)
+                {
+                    Log(LOG_LEVEL_VERBOSE, "%s: line %d from ENV file '%s' exceeded byte limit %lu, done with file",
+                        myname, linenumber, input_path, (long unsigned int)size_max);
+                    free(line);
+                    break;
+                }
+
+                if (line[0] == '#' || strlen(line) == 0)
+                {
+                    // skip comment or blank line
+                }
+                else
+                {
+                    Rlist *pieces = RlistFromSplitRegex(line, "=", 2, true);
+                    if (pieces && pieces->next)
+                    {
+                        const char* v = RlistScalarValueSafe(pieces->next);
+                        if (strlen(v) > 1 && v[0] == '"' && v[strlen(v)-1] == '"')
+                        {
+                            char* decoded = JsonDecodeString(v);
```

Comment:
It seemed like the right thing to do, based on the URL I quoted. ENV files don't support heredocs or other multiline data, as far as I know. They are not actual shell scripts, although you can source them from a shell. For instance Docker has a `--env-file FILE.env` option that doesn't know about shell variable expansion.